### PR TITLE
Deprecated serialized fields

### PR DESCRIPTION
**Pull request by VysotskiVadim**
mapbox-java used to serialized `access-token` and `uuid` when a user calls `RouteOptions#toJson()`. You will get an invalid URL If you deserialize `RouteOptions` from a json, which contains `access-token` or `uuid`, and then call `RouteOptions#toURL`.

### CODE REVIEW

**Comment by VysotskiVadim** [comment]
`multipleSetRouteCall_longAndShortRouteSameConsumerTest` failed on 1 of 3 devices. As this test tests `RouteLine`, I consider this failure a flack.

File: services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java [hunk 1]
```diff
@@ -713,6 +714,36 @@ public void routeOptionsWithUnrecognizedObjectsAndArrays() throws IOException {
...
+    Gson gson = new GsonBuilder().create();
+    JsonObject routeOptionsJson = gson.fromJson(optionsJson, JsonObject.class);
+    routeOptionsJson.addProperty("access_token", "wrong");
+    String routeOptionsJsonString = routeOptionsJson.toString();
+
+    URL url = RouteOptions.fromJson(routeOptionsJsonString).toUrl("right");
+
+    assertFalse(
+      "url shouldn't contain tokens from json: " + url,
+      url.getQuery().contains("access_token=wrong")
```
Comment (RingerJK) [inline services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java:728]:
```suggestion
      url.getQuery().contains("access_token")
```

Comment (RingerJK) [inline services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java:728]:
```suggestion
      url.getQuery().contains("access_token")
```
```suggestion
      url.getQuery().contains("access_token=wrong")
```

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java [hunk 2]
```diff
@@ -38,6 +40,12 @@ public abstract class RouteOptions extends DirectionsJsonObject {
 
   private static final String UTF_8 = "UTF-8";
   private static final String ACCESS_TOKEN_URL_PARAM_NAME = "access_token";
+  private static final Set<String> DEPRECATED_SERIALIZED_FIELDS = new HashSet<String>() {
+    {
+      add(ACCESS_TOKEN_URL_PARAM_NAME);
+      add("uuid");
+    }
+  };
```
Comment (RingerJK) [inline services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java:48]:
```suggestion
  /** 
  *  Fields that mustn't be serialized
  * - <b>access_token</b> - token must be a part of RouteOptions, it provides with URL via {@link RouteOptions#toUrl}
  * - <b>uuid</b> - legacy field, skipped for forward compatibility
  */
  private static final Set<String> SKIPPED_SERIALIZED_FIELDS =  Collections.unmodifiableSet(new HashSet<String>() {
    {
      add(ACCESS_TOKEN_URL_PARAM_NAME);
      add("uuid");
    }
  });
```

Comment (RingerJK) [inline services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java:48]:
```suggestion
  /** 
  *  Fields that mustn't be serialized
  * - <b>access_token</b> - token must be a part of RouteOptions, it provides with URL via {@link RouteOptions#toUrl}
  * - <b>uuid</b> - legacy field, skipped for forward compatibility
  */
  private static final Set<String> SKIPPED_SERIALIZED_FIELDS =  Collections.unmodifiableSet(new HashSet<String>() {
    {
      add(ACCESS_TOKEN_URL_PARAM_NAME);
      add("uuid");
    }
  });
```

**Review by RingerJK — COMMENTED** [review]
LGTM. Let's add a check (+logs) to `toUrl` that checks that we don't have the same queries twice or more time. wdyt?

**Comment by VysotskiVadim** [comment]
> Let's add a check (+logs) to toUrl that checks that we don't have the same queries twice or more time. wdyt?

@RingerJK , how do you think, what's the best think we can do if there're duplicated parameters? Throw an exceptions with an understandable description?

**Comment by RingerJK** [comment]
I think logs with `w` level should be good.

File: services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java (hunk 1 above)
Comment (VysotskiVadim) [reply services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java:728]: The URL should contain 1 `access_token`, the suggested check will fail the test.

Comment (RingerJK) [reply services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java:728]: yeah, that makes sense. Lets' also check there's exact `1` access token

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java [hunk 3]
```diff
@@ -968,6 +976,9 @@ public URL toUrl(@NonNull String accessToken) {
     if (unrecognized != null) {
       for (Map.Entry<String, SerializableJsonElement> entry : unrecognized.entrySet()) {
         JsonElement element = entry.getValue().getElement();
+        if (DEPRECATED_SERIALIZED_FIELDS.contains(entry.getKey())) {
```
[the remark this answers is not in the document]
Comment (LukasPaczos) [reply services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java:979]: LGTM

[the remark this answers is not in the document]
Comment (LukasPaczos) [reply services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java:979]: A warning log could be good.

**Review by Guardiola31337 — COMMENTED** [review]
Should we add a sanity test that serializes a legacy `DirectionsResponse` JSON like [`multiple_routes.json`](https://github.com/mapbox/mapbox-navigation-android/blob/main/instrumentation-tests/src/main/res/raw/multiple_routes.json) and verifies that the URL doesn't contain neither a duplicated `access_token` nor the `uuid`? 🤔

**Comment by VysotskiVadim** [comment]
> Should we add a sanity test that serializes a legacy `DirectionsResponse` JSON like [`multiple_routes.json`](https://github.com/mapbox/mapbox-navigation-android/blob/main/instrumentation-tests/src/main/res/raw/multiple_routes.json) and verifies that the URL doesn't contain neither a duplicated `access_token` nor the `uuid`? 🤔

added, see `sanityCheckForPreviouslySerializedRouteResponse`

**Comment by VysotskiVadim** [comment]
> I think logs with `w` level should be good.

@RingerJK , @LukasPaczos , I don't see any logging mechanisms in mapbox-java. Do you think it's time to introduce one?

TBH, I don't think that it's time 🙂

File: services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java (hunk 1 above)
Comment (VysotskiVadim) [reply services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/RouteOptionsTest.java]: added

**Comment by RingerJK** [comment]
> > I think logs with `w` level should be good.
> 
> @RingerJK , @LukasPaczos , I don't see any logging mechanisms in mapbox-java. Do you think it's time to introduce one?
> 
> TBH, I don't think that it's time 🙂

suppose, mapbox-java contains business logic, so it's time to investigate if it needs a logging mechanism. Let's cut a ticket

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java (hunk 2 above)
Comment (RingerJK) [reply services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java:55]: please put also under `Collections.unmodifiableSet`

**Comment by VysotskiVadim** [comment]
> > > I think logs with `w` level should be good.
> > 
> > 
> > @RingerJK , @LukasPaczos , I don't see any logging mechanisms in mapbox-java. Do you think it's time to introduce one?
> > TBH, I don't think that it's time 🙂
> 
> suppose, mapbox-java contains business logic, so it's time to investigate if it needs a logging mechanism. Let's cut a ticket

https://github.com/mapbox/mapbox-java/issues/1448

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java (hunk 2 above)
Comment (VysotskiVadim) [reply services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/RouteOptions.java]: thanks, added

**Comment by RingerJK** [comment]
as we still don't have logger, could you put `// TODO log me "CONTEXT"` to places that must be logged

**Comment by VysotskiVadim** [comment]
> as we still don't have logger, could you put `// TODO log.W("CONTEXT")` (replace `W` by any other level where it needs) to places that must be logged

Added to the logger ticket instead https://github.com/mapbox/mapbox-java/issues/1448#issuecomment-1147309804

**Review by RingerJK — APPROVED** [review]
LGTM

**Pull request merged**